libservo: Add initial `SiteDataManager::site_data` API

janvarga opened the pull request:
This PR adds an initial `site_data` API to `SiteDataManager`, allowing
embedders to obtain a unified list of sites that have associated site data.

The implementation introduces supporting infrastructure across storage and
`SiteDataManager` layers, including a small helper for obtaining WebStorage
origins and initial support for filtering by storage type. At this stage,
sites correspond directly to origins, future work will extend this to higher
level grouping (e.g. domain or eTLD+1).

Testing: Added a new integration test covering site data behavior across
localStorage and sessionStorage.

janvarga committed:
libservo: Add storage threads to `SiteDataManager`

This patch adds access to storage threads to `SiteDataManager`, mirroring the
existing handling of resource threads.

This is preparatory work for invoking storage related operations from
`SiteDataManager`.

Signed-off-by: Jan Varga <jvarga@igalia.com>

janvarga committed:
libservo: Rename `site_data.rs` to `site_data_manager.rs` in tests

This patch renames `site_data.rs` to `site_data_manager.rs` in tests for
consistency with `SiteDataManager` rust file.

Signed-off-by: Jan Varga <jvarga@igalia.com>

janvarga committed:
storage: Rename OriginDescriptors message to ListOrigins

This patch renames the `OriginDescriptors` message to `ListOrigins`, which
better describes its purpose as returning a list of available origins.

Signed-off-by: Jan Varga <jvarga@igalia.com>

janvarga committed:
storage: Add `StorageThreads::list_webstorage_origins` helper

This patch adds a helper `StorageThreads::list_webstorage_origins`, mirroring
existing helpers such as `ResourceThreads::clear_cookies`.

The new method provides a simple way to list available WebStorage origins
without exposing lower level messaging details.

Signed-off-by: Jan Varga <jvarga@igalia.com>

janvarga committed:
libservo: Add `SiteDataManager::list_sites` with storage type filtering

This patch adds a `list_sites` API to `SiteDataManager`, along with initial
infrastructure for specifying which storage types should be considered.

At this stage, sites correspond directly to storage origins. In the future,
this will be extended to support higher level grouping (for example by
eTLD+1).

This functionality is intended to provide a unified view of site data for
embedders.

Signed-off-by: Jan Varga <jvarga@igalia.com>

janvarga committed:
libservo: Expose storage types in `SiteData`

This patch extends `SiteData` to include information about which storage types
are present for each site.

Signed-off-by: Jan Varga <jvarga@igalia.com>

mrobinson commented on components/servo/tests/site_data_manager.rs:
```diff
@@ -0,0 +1,207 @@
+        ]
+    };
+}
+
```
Would you be willing to turn this macro a function to turn a vector of tuples into structs? Performance isn't critical here and macros make it difficult to follow the code in the IDE.

mrobinson commented on components/servo/site_data_manager.rs:
```diff
@@ -2,22 +2,97 @@
+        const ALL =
+                   Self::LOCAL_STORAGE.bits() |
+                   Self::SESSION_STORAGE.bits();
+    }
```
`bitflags` already generates a `fn all()`, so this is unnecessary. See https://docs.rs/bitflags/latest/bitflags/example_generated/struct.Flags.html#method.all.

mrobinson commented on components/servo/site_data_manager.rs:
```diff
@@ -2,22 +2,97 @@
+    #[derive(Clone, Copy, Debug, PartialEq)]
+    pub struct StorageTypes: u32 {
+        const LOCAL_STORAGE   = 1 << 0;
+        const SESSION_STORAGE = 1 << 1;
```
Do you mind adding some rustdoc, perhaps with links to the API either in MDN or the specification to explain what these values mean?

```suggestion
        const Local   = 1 << 0;
        const Session = 1 << 1;
```

mrobinson commented on components/servo/site_data_manager.rs:
```diff
@@ -2,22 +2,97 @@
+        }
+    }
+
+    pub fn list_sites(&self, storage_types: StorageTypes) -> Vec<SiteData> {
```
This is a public API so could probably use a bit of rustdoc.

mrobinson commented on components/shared/storage/lib.rs:
```diff
@@ -33,6 +33,14 @@ impl StorageThreads {
         }
     }
+
+    pub fn list_webstorage_origins(&self, storage_type: StorageType) -> Vec<OriginDescriptor> {
```
```suggestion
    pub fn webstorage_origins(&self, storage_type: StorageType) -> Vec<OriginDescriptor> {
```

mrobinson commented on components/storage/webstorage/mod.rs:
```diff
@@ -248,8 +248,9 @@ impl WebStorageManager {
-                    self.origin_descriptors(sender, storage_type);
+                WebStorageThreadMsg::ListOrigins(sender, storage_type) => {
+                    let origin_descriptors = self.origin_descriptors(storage_type);
+                    let _ = sender.send(origin_descriptors);
```
Nit: no need for a temporary here:

```rust
                    let _ = sender.send( self.origin_descriptors(storage_type));
```

mrobinson commented on components/storage/webstorage/mod.rs:
```diff
@@ -466,6 +467,9 @@ impl WebStorageManager {
         name: String,
         value: String,
     ) {
+        // XXX localStorage and sessionStorage shouldn't share usage and quota.
```
```suggestion
        // TODO: localStorage and sessionStorage shouldn't share usage and quota.
```

mrobinson reviewed (changes requested):
Looks good. I have a few small comments below. I would get rid of the `list_` terminology. Rust is typically a bit terser.

janvarga commented:
> Looks good. I have a few small comments below. I would get rid of the `list_` terminology. Rust is typically a bit terser.

Thanks for the comments. It all makes sense. Let me incorporate the suggested changes.

janvarga committed:
libservo: Address review feedback for `SiteDataManager` listing support

This patch incorporates review feedback by performing the following cleanups
and refinements:

- Rename `list_webstorage_origins` to `webstorage_origins`
- Rename `StorageTypes` to `StorageType`
- Rename `list_sites` to `site_data`
- Replace `XXX` comments with `TODO`
- Inline `SiteDataBuilder` logic into `site_data`
- Remove the `site_data_vec` test helper macro
- Add rustdoc for public structs and APIs

Signed-off-by: Jan Varga <jvarga@igalia.com>

janvarga committed:
storage: Rename StorageType enum for consistency

This patch performs a remaining round of renaming in the web storage code to
to avoid name conflicts with the newly introduced `StorageType` struct used by
`SiteDataManager`.

The change is a follow-up to PR #39550 and PR #40090 and does not introduce
any functional changes..

Signed-off-by: Jan Varga <jvarga@igalia.com>

janvarga commented on components/servo/site_data_manager.rs:
```diff
@@ -2,22 +2,97 @@
+
+bitflags! {
+    #[derive(Clone, Copy, Debug, PartialEq)]
+    pub struct StorageTypes: u32 {
```
`u8` should be enough for now.

janvarga commented on components/servo/site_data_manager.rs:
```diff
@@ -2,22 +2,97 @@
 use net_traits::ResourceThreads;
+use rustc_hash::FxHashMap;
+use storage_traits::StorageThreads;
+use storage_traits::webstorage_thread::{OriginDescriptor, StorageType};
```
Hm, now I remember that I actually already wanted to rename `storage_traits::webstorage_thread::StorageType` to `storage_traits::webstorage_thread::WebStorageType`
I added a new commit for that (hope that's ok).

janvarga commented on components/servo/site_data_manager.rs:
```diff
@@ -30,3 +105,31 @@ impl SiteDataManager {
+                .and_modify(|types| *types |= storage_type)
+                .or_insert(storage_type);
+        }
+    }
```
This was a preparation for the grouping mentioned in commit messages and in the PR description, but I agree, let's add a builder later once it's actually needed.

mrobinson reviewed (approved):
Looks great with one tiny nit:
